model/inline/src/main/scala/aqua/model/inline/ArrowInliner.scala

```diff
@@ -373,6 +371,56 @@ object ArrowInliner extends Logging {
     } yield values -> arrows
   }
 
+  //
+  private def collectionsToCanons(
+    tree: RawTag.Tree,
+    streamArgs: Map[String, VarModel]
+  ): RawTag.Tree = {
+    // collect arguments with stream type
+    // to exclude it from resolving and rename it with a higher-level stream that passed by argument
+    val streamsToRename = streamArgs.view.mapValues(_.name).toMap
+
+    if (streamsToRename.isEmpty) tree
+    else
+      tree
+        .map(_.mapValues(_.map {
+          // if an argument is a BoxType (Array or Option), but we pass a stream,
+          // change a type as stream to not miss `$` sign in air
+          case v @ VarRaw(name, baseType: CollectionType) if streamsToRename.contains(name) =>
+            v.copy(baseType = CanonStreamType(baseType.element))
```

Comment:
Frankly speaking, I don't like this type manipulation, it is one more non-local thing we should remember. Can we add one `ap` to `canon` to make honest array from passed stream? And we will optimize it later if it will prove beneficial.